sdk/core/azure-core/src/main/java/com/azure/core/implementation/ReliableDownloadFlux.java

```diff
@@ -0,0 +1,58 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import reactor.core.CoreSubscriber;
+import reactor.core.publisher.Flux;
+
+import java.nio.ByteBuffer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * A {@code Flux<ByteBuffer>} implementation which is capable of performing a reliable download by applying a resume
+ * operation if an error occurs during the download.
+ */
+public final class ReliableDownloadFlux extends Flux<ByteBuffer> {
+    private final Supplier<Flux<ByteBuffer>> initialDownloadSupplier;
+    private final Predicate<Throwable> resumePredicate;
+    private final int maxRetries;
+    private final Function<Long, Flux<ByteBuffer>> resumeDownload;
```

Comment:
For storage cases. It would be great to somehow pass ETag acquired in initial step to the resume part. Otherwise we're risking data corruption if underlying resource has been changed meanwhile.

Reply:
maybe something like:
```
Supplier<Tupple<TInitialState, Flux<ByteBuffer>>> initialDownloder;
Function<Long, TInitialState, Flux<ByteBuffer>>> resumeDownload;
```
So that service can smuggle some properties from intial call to resume call by providing own TInitialState shape.

Reply:
Wouldn't this state be a closure capture into the `resumeDownload` lambda function? It appears that is how it is done for .NET: https://github.com/Azure/azure-sdk-for-net/blob/1f1e21a699cb82acc45f9b566a5e1c10a0f36720/sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs#L1044

Reply:
Here is an example of using it in download: https://github.com/Azure/azure-sdk-for-java/pull/21838

Reply:
Thank you, looks good.